crates/containerd-shim-wasm/src/sandbox/shim/otel.rs

```diff
@@ -0,0 +1,164 @@
+//! OpenTelemetry Configuration Module
+//!
+//! This module provides a configuration structure and associated methods to initialize
+//! OpenTelemetry tracing with the OTLP exporter. The configuration can be set up via
+//! the `OtelConfig` struct and its builder pattern.
+//!
+//! # Usage
+//!
+//! ```rust
+//! use containerd-shim-wasm::sandbox::shim::otel::{OtelConfig, OTEL_EXPORTER_OTLP_ENDPOINT};
+//!
+//! fn main() -> anyhow::Result<()> {
+//!     let otel_endpoint = std::env::var(OTEL_EXPORTER_OTLP_ENDPOINT).expect("OTEL_EXPORTER_OTLP_ENDPOINT not set");
+//!     let otel_config = OtelConfig::builder()
+//!         .otel_endpoint(otel_endpoint)
+//!         .name("my-service".to_string())
+//!         .build()?;
+//!
+//!     let _guard = otel_config.init()?;
+//!
+//!     // Your application code here
+//!
+//!     Ok(())
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use opentelemetry::global::{self, set_text_map_propagator};
+use opentelemetry::trace::TraceError;
+use opentelemetry_otlp::{
+    SpanExporterBuilder, WithExportConfig, OTEL_EXPORTER_OTLP_PROTOCOL_DEFAULT,
+};
+pub use opentelemetry_otlp::{OTEL_EXPORTER_OTLP_ENDPOINT, OTEL_EXPORTER_OTLP_PROTOCOL};
+use opentelemetry_sdk::propagation::TraceContextPropagator;
+use opentelemetry_sdk::{runtime, trace as sdktrace};
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt as _;
+use tracing_subscriber::layer::SubscriberExt as _;
+use tracing_subscriber::{EnvFilter, Registry};
+
+const OTEL_EXPORTER_OTLP_PROTOCOL_HTTP_PROTOBUF: &str = "http/protobuf";
+const OTEL_EXPORTER_OTLP_PROTOCOL_GRPC: &str = "grpc";
```

Comment:
As you've mentioned these two `String` constants are only used in L.108 so I don't see the value to define an extra enum for this. They are also used for string comparison with the value of environment variables to determine what protocol we are using.

Reply:
Well, they are not used only in L.108, they are used by whoever calls `pub fn otel_protocol(mut self, otel_protocol: String) -> Self;`. Because they are strings, you need an error condition there
```rust
            _ => Err(TraceError::from(
                "Invalid OTEL_EXPORTER_OTLP_PROTOCOL value",
            ))?,
```
If it was an enum, it makes it cleas to consumers as to what are the valid options.

Although I also see your point that this is coming from an env var, and using an enum would defer handling the string matching to the consumer.

What do you think of having that method `OtelConfig::build_from_env() -> Result<OtelConfig>`? that parses the standard env-vars and builds the config acordingly. That would let us encapsulate all the otel logic in this module, and that could also be in charge of doing the string parcing.

Reply:
Okay, I am actually going to delete the builder and expose `build_from_env` from `Config`.